web/handlers.go

```diff
@@ -102,10 +102,29 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	if len(token) != 0 {
 		session, err := c.App.GetSession(token)
+		if err != nil {
+			c.Log.Info("Invalid session", mlog.Err(err))
+			if err.StatusCode == http.StatusInternalServerError {
+				c.Err = err
+			} else if h.RequireSession {
+				c.RemoveSessionCookie(w, r)
+				c.Err = model.NewAppError("ServeHTTP", "api.context.session_expired.app_error", nil, "token="+token, http.StatusUnauthorized)
+			}
+		} else if !session.IsOAuth && tokenLocation == app.TokenLocationQueryString {
+			c.Err = model.NewAppError("ServeHTTP", "api.context.token_provided.app_error", nil, "token="+token, http.StatusUnauthorized)
+		} else {
+			c.App.Session = *session
+		}
+
+		// Rate limit by UserID
+		if c.App.Srv.RateLimiter != nil && c.App.Srv.RateLimiter.UserIdRateLimit(c.App.Session.UserId, w) {
+			return
+		}
+
 		csrfCheckPassed := false
 
 		// CSRF Check
-		if tokenLocation == app.TokenLocationCookie && h.RequireSession && !h.TrustRequester && r.Method != "GET" {
+		if c.Err == nil && tokenLocation == app.TokenLocationCookie && h.RequireSession && !h.TrustRequester && r.Method != "GET" {
```

Comment:
Is there any problem with allowing validation of the session to happen for CSRF attempts? Can any info be gained?
Would it be simpler to just skip CSRF if the session can't be retrieved?

Reply:
If the session cannot be validated, we won't validate the CSRF since `c.Err != nil`. Without a valid session, we cannot validate the CSRF because it's bound to the session itself. 

Since CSRF depends on Session, but Session not on CSRF, the order makes sense for me, especially moving the rate limit before it. Additionally, for CSRF attack no information is sent back from Mattermost to the attackers side, so information disclosure is no risk here either.